samples/hello/scripts/generate-keys.py

```diff
@@ -6,43 +6,42 @@
 import asyncclick as click
 from golioth import Client
 
-
-SCRIPTS_DIR = Path(__file__).resolve().parents[0]
-SAMPLE_DIR = SCRIPTS_DIR.parents[0]
-KEYS_DIR = SAMPLE_DIR / 'keys'
-
 ROOT_CA_NAME = 'root-ca'
 
-
 def run(*args, **kwargs):
     if 'check' not in kwargs:
         kwargs['check'] = True
 
     if 'shell' not in kwargs:
         kwargs['shell'] = True
 
-    print(f'Running: {args}')
     return subprocess.run(*args, **kwargs)
 
+@click.group()
+async def cli():
+    pass
 
 @click.command()
+@click.option('--upload/--no-upload', default=False, help='Upload root certificate to Golioth')
 @click.option('-c', '--config-path', type=Path,
               help='Path to goliothctl configuration')
 @click.option('--api-key', help='API key')
-@click.option('-d', '--device-name',
-              help='Name of device on which RPC method will be called',
-              required=True)
-async def main(config_path, api_key, device_name):
-    """Generate certificates, add them to Golioth and populate to keys/ directory in sample."""
-    client = Client(config_path, api_key=api_key)
-    project = await client.default_project()
-    device = await project.device_by_name(device_name)
+@click.option('--project', 'project_name', help='Project where root certificate will be uploaded')
+@click.option('-o', '--output', type=Path, default=Path('.'), help="Output directory")
+async def root(upload, config_path, api_key, project_name, output):
+    if upload:
+        if not api_key:
+            click.echo("API key required to upload root certificate")
+
+        client = Client(config_path, api_key=api_key)
 
-    primary_hardware_id = device.info['hardwareIds'][0]
-    client_name = f"{project.id}-{device.name}"
+        if project_name:
+            project = await client.project_by_name(project_name)
+        else:
+            project = await client.default_project()
 
-    root_ca_crt = SCRIPTS_DIR / f'{ROOT_CA_NAME}-{project.id}.crt.pem'
-    root_ca_key = SCRIPTS_DIR / f'{ROOT_CA_NAME}-{project.id}.key.pem'
+    root_ca_crt = output / f'{ROOT_CA_NAME}-{project_name}.crt.pem'
+    root_ca_key = output / f'{ROOT_CA_NAME}-{project_name}.key.pem'
```

Comment:
In the first function there is `project_name` and in the second there is `project_id`. Also, in the second function argument description there is "Project slug". It is all very confusing.

Reply:
The `project_name` and `project_id` are actually different. In the root certificate generation command, we already need the API key and goliothctl config path if we're doing in the upload step - in that case, we can use goliothctl to get the project-id by name. If we're not doing the upload in the script (i.e. we'll upload the root cert manually later), then we don't need the project name or project id. For the device generation command, there's no communication necessary with the Golioth servers (this is actually one of the main reason I split them into separate commands). In that case, you need to provide the project-id because we have no way of converting project-name to project-id.

I'll change the root generation command to accept project-id instead of project-name. It will be a little less convenient for the user, but I think the benefit from reducing confusion is greater. Thanks for the feedback!